x/ratelimit/rate_limit.go

```diff
@@ -0,0 +1,155 @@
+//
+// For NATIVE denoms, return as is (e.g. ustrd)
+// For NON-NATIVE denoms, take the ibc hash (e.g. hash "transfer/channel-2/usoms" into "ibc/...")
+func ParseDenomFromSendPacket(packet transfertypes.FungibleTokenPacketData) (denom string) {
+	// Determine the denom by looking at the denom trace path
+	denomTrace := transfertypes.ParseDenomTrace(packet.Denom)
+
+	// Native assets will have an empty trace path and can be returned as is
+	if denomTrace.Path == "" {
+		denom = packet.Denom
+	} else {
+		// Non-native assets should be hashed
+		denom = denomTrace.IBCDenom()
+	}
+
+	return denom
+}
+
+// Parse the denom from the Recv Packet that will be used by the rate limit module
+// The denom that the rate limiter will use for a RECEIVE packet depends on whether it was a source or sink
+// 		Source: The packet is being received by a chain it was just sent from (i.e. the token has gone back and forth)
+//              (e.g. strd is sent -> to osmosis -> and then back to stride)
+//      Sink:   The packet is being received by a chain that either created it or previous received it from somewhere else
+//              (e.g. atom is sent -> to stride) (e.g.2. atom is sent -> to osmosis -> which is then sent to stride)
+//
+//      If the chain is acting as a SINK:
+//      	We add on the Stride port and channel and hash it
+//          Ex1: uosmo sent from Osmosis to Stride
+//              Packet Denom:   uosmo
+//               -> Add Prefix: transfer/channel-X/uosmo
+//               -> Hash:       ibc/...
+//
+//          Ex2: ujuno sent from Osmosis to Stride
+//              PacketDenom:    transfer/channel-Y/ujuno  (channel-Y is the Juno <> Osmosis channel)
+//               -> Add Prefix: transfer/channel-X/transfer/channel-Y/ujuno
+//               -> Hash:       ibc/...
+//
+//      If the chain is acting as a SOURCE:
+//      	First, remove the prefix. Then if there is still a denom trace, hash it
+//          Ex1: ustrd sent back to Stride from Osmosis
+//              Packet Denom:      transfer/channel-X/ustrd
+//               -> Remove Prefix: ustrd
+//               -> Leave as is:   ustrd
+//
+//			Ex2: juno was sent to Stride, then to Osmosis, then back to Stride
+//              Packet Denom:      transfer/channel-X/transfer/channel-Z/ujuno
+//               -> Remove Prefix: transfer/channel-Z/ujuno
+//               -> Hash:          ibc/...
+func ParseDenomFromRecvPacket(packet channeltypes.Packet, packetData transfertypes.FungibleTokenPacketData) (denom string) {
+	// To determine the denom, first check whether Stride is acting as source
+	if transfertypes.ReceiverChainIsSource(packet.GetSourcePort(), packet.GetSourceChannel(), packetData.Denom) {
+		// Remove the source prefix (e.g. transfer/channel-X/transfer/channel-Z/ujuno -> transfer/channel-Z/ujuno)
+		sourcePrefix := transfertypes.GetDenomPrefix(packet.GetSourcePort(), packet.GetSourceChannel())
```

Comment:
Is this logic equivalent to what ICS-20 does when it unwraps vouchers? If so, let's make sure we pull the code from there and reference it

Reply:
Yeah I used the ibc-go code as a reference
https://github.com/cosmos/ibc-go/blob/main/modules/apps/transfer/keeper/relay.go#L200